### PR TITLE
Binning bug

### PR DESCRIPTION
**Pull request by zainsohail04**
Just setting up PR for #95 so we can figure this out. I came with a rough fix that shouldn't be used.

Note: Build will fail since test is testing num_steps.

### CODE REVIEW

**Comment by rettigl** [comment]
I implemented now a version close to what I discussed in #95. I did not include the extension of the number of bins by one, but this can be easily added. What it does not is the following:
- For ranges, bins, it returns bin centers, aligned at (left range, ... right range except last bin).
Example: (0, 10), 10 gives [0, 1, 2, 3, 4, 5, 6, 7, 8, 9] as bin _centers_.
- For bins as arrays, they are interpreted as bin centers, and converted into edges before binning.
I also cleaned up the simplify_binning_arguments function, and wrote a test function covering all cases. I removed the case that bins can be a single tuple of bins and ranges, because I don't see a purpose for it (if you have multiple axes, they will most certainly never have the same ranges...).

**Review by steinnymir — APPROVED** [review]
Again the same problem... 40 files changed.
This PR is about the bug in the binning function, so why does it refactor almost all docstrings of all functions? are these really changes here or do they come from merging other PRs in here?

The code changes I was able to find looked good. But I am certain I missed some.

For the sake of moving on, I will accept this. But lets try keep bug fix PRs to the point.

comments:
- I agree with the removal of the tuple method in the binning arguments. I added far too many options, but that is the default from numpy...
- for other definitions, we can continue our discussion in #95 and make smaller, targeted PRs eventually.

**Comment by rettigl** [comment]
> Again the same problem... 40 files changed.

My apologies, this is a mistake in rebasing/merging of upstream I apparently made, which overwrites a lot of recent changes in master. Thanks for spotting this. Rebasing and force pushing should solve this, I believe.

**Comment by rettigl** [comment]
Now it is properly rebased, I believe.

File: tests/test_binning.py [hunk 1]
```diff
@@ -199,44 +201,105 @@ def test_bin_edges_to_bin_centers():
...
-def test_simplify_binning_arguments_tuple():
-    """Test the result of the _simplify_binning_arguments functions for bin tuples"""
-    bin_tuple = [tuple((r[0], r[1], b)) for r, b in zip(ranges, bins)]
-    bins_, axes_, ranges_ = _simplify_binning_arguments(bin_tuple, axes)
-    assert bins_ == bins
-    assert axes_ == axes
-    assert ranges_ == ranges
+    #    for (bin_, bin__) in zip(bins__, bin_centers):
+    #        np.testing.assert_array_equal(bin_, bin__)
+    assert bins__ == bins_expected
```
Comment (steinnymir) [inline tests/test_binning.py:303]: why not use `np.testing.assert_allclose` here? or if they need to be identical: `np.testing.assert_array_equal`

**Review by steinnymir — APPROVED** [review]
Nice! now I managed to read through easily! 😄 
looks good to me!

File: tests/test_binning.py (hunk 1 above)
Comment (rettigl) [reply tests/test_binning.py:303]: They are not arrays, but lists of int or np.array at that point. Not sure if that will work, but I can try. Also not sure what the == actually tests, though.

**Comment by steinnymir** [comment]
I think this is a good definition. It kind of sounds pythonic, when you think of how `range` works for example. We could have this as default behaviour, and add the endpoint argument to it allowing to include the last point when really necessary.
I believe there will be very few cases in which this would be used, but almost certainly someone who will complain if it is not there...

File: tests/test_binning.py (hunk 1 above)
Comment (rettigl) [reply tests/test_binning.py]: Updated to assert_array_equals now

**Pull request merged**